common/types/ballot.go

```diff
@@ -106,16 +110,37 @@ type VotingEligibilityProof struct {
 	Sig []byte
 }
 
+// Initialize calculates and sets the Ballot's cached ballotID and smesherID.
+// this should be called once all the other fields of the Ballot are set.
+func (b *Ballot) Initialize() error {
+	if b.ID() != EmptyBallotID {
+		return fmt.Errorf("ballot already initialized")
+	}
+
+	bytes := b.Bytes()
+	b.ballotID = BallotID(CalcHash32(bytes).ToHash20())
+	pubkey, err := ed25519.ExtractPublicKey(bytes, b.Signature)
+	if err != nil {
+		return fmt.Errorf("ballot initialize: %w", err)
```

Comment:
should be about extracting pubkey from sig?

Reply:
done